### PR TITLE
[sqlserver] Be consistent with the connection string generation.

### PR DESCRIPTION
**Pull request by truthbk**

### CODE REVIEW

**Comment by olivielpeau** [comment]
Let's fix the logic of `open_db_connections` before merging this.

**Comment by Martiflex** [comment]
@truthbk FYI Reports of error with `custom_metrics` activated for this modified check. More info https://datadog.zendesk.com/agent/tickets/47559

File: checks.d/sqlserver.py [hunk 1]
```diff
@@ -307,34 +307,42 @@ def close_cursor(self, cursor):
...
-                connection['conn'].close()
-            except Exception as e:
-                self.log.warning("Could not close adodbapi db connection\n{0}".format(e))
+        conn_key = self._conn_key(instance)
+        if conn_key not in self.connections:
+            return
 
-    def open_db_connections(self):
+        try:
+            self.connections[conn_key]['conn'].close
```
Comment (olivielpeau) [inline checks.d/sqlserver.py]: `close` -> `close()`  (probably)

**Comment by olivielpeau** [comment]
Looks good overall!

As an aside, I think we could make the whole connection handling logic way clearer if we used context managers, i.e.:
- a `get_connection(instance)` context manager that would return the connection for the instance and take care of opening and closing it when used with `with get_connection(instance) as connection:`
- a `get_cursor(connection)` that would be passed the instance's connection (since the current `get_cursor` only ever uses an instance's connection anyway) and return a cursor. `adodbapi`'s `cursor` can already be used as a context manager that closes the cursor automagically.

Probably outside the scope of this PR, but the logic is so convoluted right now that it'd be worth spending some time on it at some point.

**Comment by truthbk** [comment]
Totally agree, we should definitely rewrite this.

File: checks.d/sqlserver.py (hunk 1 above)
Comment (truthbk) [inline checks.d/sqlserver.py]: indeedy... typo!

**Comment by olivielpeau** [comment]
:+1: Feel free to squash and merge! :tada:

File: checks.d/sqlserver.py [hunk 2]
```diff
@@ -307,34 +307,42 @@ def close_cursor(self, cursor):
...
+        timeout = int(instance.get('command_timeout',
+                                   self.DEFAULT_COMMAND_TIMEOUT))
+        try:
+            rawconn = adodbapi.connect(self._conn_string(instance=instance),
+                                    timeout=timeout)
+            if conn_key not in self.connections:
+                self.connections[conn_key] = {'conn': rawconn, 'timeout': timeout}
+            else:
+                # Existing Connection object will be closed (if open) when GC'd
+                self.connections[conn_key]['conn'] = rawconn
```
Comment (olivielpeau) [inline checks.d/sqlserver.py]: I don't really know how adodbapi's connections behave exactly, but given that we used to make sure to always close the connections explicitely (and not wait for GC to take care of it), we should probably explicitely close the old connection here before replacing it with the new one.

**Comment by olivielpeau** [comment]
Added a comment, just want to make extra-sure that we won't have stale open connections as we seem to have had trouble with this in the past.

File: checks.d/sqlserver.py (hunk 2 above)
Comment (truthbk) [inline checks.d/sqlserver.py]: They call close() on __del__(), but I guess we can make it explicit.

**Comment by olivielpeau** [comment]
:+1:

Can you squash your commits? Feel free to merge once that's done.

**Pull request merged**